### ISSUE TITLE
Feature Request: Generate config without home-manager

### ISSUE THREAD

**Issue by minego** [sodiboo/niri-flake#278]
I'd really like to be able to use this module to generate my niri config, but I don't use home manager... I currently am doing this with: ```
				nixpkgs.overlays = [
					niri.overlays.niri

					(_final: _prev: {
						# Modify the niri startup command to point it to the
						# config file
						niri-unstable = _prev.niri-unstable.overrideAttrs (final: prev: {
							postFixup = ''
								${prev.postFixup}

								substituteInPlace $out/share/systemd/user/niri.service --replace "--session" "--session -c ${./niri.kdl}"
								'';
						});
					})
				];

				programs.niri.enable			= true;
				programs.niri.package			= pkgs.niri-unstable;
```

I'd like to be able to use this flake to generate my `niri.kdl` file though. That would make it a lot easier to make tweaks for specific machines, etc.

Is it possible to do that, and then use the output in my overlay to point to that derivation?

**Comment by sodiboo** [comment]
What exactly are you asking for?

From your description, i'm gathering you want to generate a config without using home-manager, as in the `program.niri.settings` parts? If so, you can use [the internal settings module](https://github.com/sodiboo/niri-flake/blob/a1bdc17ecfd08980b6a621fb69413977b83ebdb0/flake.nix#L328) which is exposed through `lib.internal`. you can see how this is done [here in my config](https://github.com/sodiboo/nix-config/blob/main/login.mod.nix#L23-L25) and the other half of that to genrate the file itself [here in my config](https://github.com/sodiboo/nix-config/blob/8ca2b21c61a4f23052c67e52276bb673a574e17c/login.mod.nix#L66).

From your snippet, it looks more like you want to just make it point to an arbitrary location, in which case there isn't really an easy way to do it currently. I've been thinking of implementing an environment variable in niri that lets us (in nixos land) set it in a wrapper. The current `-c` flag isn't reliable because it should still be overrideable with the prefix (it isn't) and also it preents you from using any subcommands. But i'll see what i can do on the other side of this to make this work well for you.

**Comment by minego** [comment]
I'm not sure how I messed up the formatting, but I think you understood my meaning, and the answer was perfect!

I now have it working, using -c to point to my config. The PR you put up looks like a very nice improvement too! Thank you!

**Comment by sodiboo** [comment]
I'm going to reopen this because I want to create a better interface to perform this. Yes, you can pass `--config` manually, but that's annoying if you use it more than once.

**Comment by minego** [comment]
> I'm going to reopen this because I want to create a better interface to perform this. Yes, you can pass `--config` manually, but that's annoying if you use it more than once.

Makes sense. I do have my setup working now, but I agree that it could be done in a much better way.

I appreciate this kind of dedication a lot. This kind of response makes me feel very good about using this project!

**Comment by heisfer** [comment]
Sorry for issue bump,
Reading from niri wiki, it can read config from `/etc/niri/config.kdl`. Maybe nixos-module should generate there?

**Comment by sodiboo** [comment]
> Sorry for issue bump,
> Reading from niri wiki, it can read config from `/etc/niri/config.kdl`. Maybe nixos-module should generate there?

Perhaps.

I don't actually see much utility in this, since it only defines a "default config" that can be replaced entirely by one in the home directory; it's not like, a systemwide "base config". So, I won't be adding that functionality directly to the NixOS module. But I do want to make it easier to generate a niri config, in general, not tied to NixOS or home-manager, which if you wanted to place in `environment.etc` you easily could.
